File: Autonomous Mobile Robots Course/Chapter11/Lesson1/Chapter11_Lesson1.py
# Chapter11_Lesson1.py
"""
Autonomous Mobile Robots (Control Engineering)
Chapter 11: SLAM I — Filter-Based SLAM
Lesson 1: The SLAM Problem Formulation

This script implements a *minimal* filter-based SLAM core using an EKF over the
augmented state y = [x, y, theta, m1x, m1y, ..., mNx, mNy].

Pedagogical choices (for Lesson 1):
- Known data association (each measurement is already matched to a landmark id).
- Simple differential-drive motion model (unicycle).
- Range-bearing landmark observations.
- Gaussian noise; standard EKF predict-update.

Note: EKF-SLAM structure/limitations are treated in Lesson 2; here we focus on
the probabilistic formulation and the recursion implemented by a filter.
"""

from __future__ import annotations
import numpy as np


def wrap_angle(a: float) -> float:
    """Wrap angle to (-pi, pi]."""
    return (a + np.pi) % (2.0 * np.pi) - np.pi


def motion_model(x: np.ndarray, u: np.ndarray) -> np.ndarray:
    """
    Unicycle motion model.
    State: x = [px, py, theta]
    Control: u = [v, omega, dt]
    """
    px, py, th = x
    v, w, dt = u
    if abs(w) < 1e-12:
        px2 = px + v * dt * np.cos(th)
        py2 = py + v * dt * np.sin(th)
        th2 = th
    else:
        px2 = px + (v / w) * (np.sin(th + w * dt) - np.sin(th))
        py2 = py - (v / w) * (np.cos(th + w * dt) - np.cos(th))
        th2 = th + w * dt
    return np.array([px2, py2, wrap_angle(th2)], dtype=float)


def jacobian_motion_wrt_state(x: np.ndarray, u: np.ndarray) -> np.ndarray:
    """
    Jacobian Fx = d f(x,u) / d x  (3x3).
    """
    px, py, th = x
    v, w, dt = u
    Fx = np.eye(3, dtype=float)

    if abs(w) < 1e-12:
        Fx[0, 2] = -v * dt * np.sin(th)
        Fx[1, 2] =  v * dt * np.cos(th)
    else:
        th2 = th + w * dt
        Fx[0, 2] = (v / w) * (np.cos(th2) - np.cos(th))
        Fx[1, 2] = (v / w) * (np.sin(th2) - np.sin(th))
    return Fx


def measurement_model_pose_landmark(x: np.ndarray, mi: np.ndarray) -> np.ndarray:
    """
    Range-bearing measurement to landmark i.
    z = [range, bearing], bearing is relative to robot heading.
    """
    px, py, th = x
    mx, my = mi
    dx = mx - px
    dy = my - py
    r = np.sqrt(dx * dx + dy * dy)
    b = np.arctan2(dy, dx) - th
    return np.array([r, wrap_angle(b)], dtype=float)


def jacobian_measurement_wrt_pose_and_landmark(x: np.ndarray, mi: np.ndarray) -> tuple[np.ndarray, np.ndarray]:
    """
    Returns (H_x, H_m) where:
    H_x = d h / d x  (2x3)
    H_m = d h / d m_i (2x2)
    """
    px, py, th = x
    mx, my = mi
    dx = mx - px
    dy = my - py
    q = dx * dx + dy * dy
    r = np.sqrt(q)

    # Robustness against division by zero
    if q < 1e-12:
        q = 1e-12
        r = np.sqrt(q)

    # dr/dx = -dx/r, dr/dy = -dy/r, dr/dtheta = 0
    # dphi/dx =  dy/q, dphi/dy = -dx/q, dphi/dtheta = -1
    Hx = np.array([
        [-dx / r, -dy / r, 0.0],
        [ dy / q, -dx / q, -1.0]
    ], dtype=float)

    # dr/dmx = dx/r, dr/dmy = dy/r
    # dphi/dmx = -dy/q, dphi/dmy = dx/q
    Hm = np.array([
        [ dx / r,  dy / r],
        [-dy / q,  dx / q]
    ], dtype=float)
    return Hx, Hm


def ekf_slam_predict(y: np.ndarray, P: np.ndarray, u: np.ndarray, Q_pose: np.ndarray) -> tuple[np.ndarray, np.ndarray]:
    """
    EKF predict step on augmented state.
    Q_pose: 3x3 process noise covariance on pose (in state coordinates).
    """
    n = y.size
    assert P.shape == (n, n)
    assert Q_pose.shape == (3, 3)

    x = y[:3].copy()
    m = y[3:].copy()

    Fx = jacobian_motion_wrt_state(x, u)
    x_pred = motion_model(x, u)

    # Augmented transition Jacobian
    F = np.eye(n, dtype=float)
    F[:3, :3] = Fx

    # Augmented process noise (pose only)
    Q = np.zeros((n, n), dtype=float)
    Q[:3, :3] = Q_pose

    y_pred = np.concatenate([x_pred, m])
    P_pred = F @ P @ F.T + Q
    return y_pred, P_pred


def ekf_slam_update_landmark(
    y: np.ndarray,
    P: np.ndarray,
    z: np.ndarray,
    landmark_id: int,
    R_meas: np.ndarray
) -> tuple[np.ndarray, np.ndarray, np.ndarray]:
    """
    EKF update for a single landmark observation with known data association.

    landmark_id: 0-based index into landmark list.
    z: [range, bearing]
    R_meas: 2x2 measurement noise covariance
    Returns: (y_upd, P_upd, innovation)
    """
    n = y.size
    assert P.shape == (n, n)
    assert z.shape == (2,)
    assert R_meas.shape == (2, 2)

    x = y[:3]
    i = landmark_id
    idx = 3 + 2 * i
    mi = y[idx:idx + 2]

    z_hat = measurement_model_pose_landmark(x, mi)
    innov = np.array([z[0] - z_hat[0], wrap_angle(z[1] - z_hat[1])], dtype=float)

    Hx, Hm = jacobian_measurement_wrt_pose_and_landmark(x, mi)

    H = np.zeros((2, n), dtype=float)
    H[:, :3] = Hx
    H[:, idx:idx + 2] = Hm

    S = H @ P @ H.T + R_meas
    K = P @ H.T @ np.linalg.inv(S)

    y_upd = y + K @ innov
    y_upd[2] = wrap_angle(y_upd[2])

    # Joseph form for covariance update (numerically safer)
    I = np.eye(n, dtype=float)
    P_upd = (I - K @ H) @ P @ (I - K @ H).T + K @ R_meas @ K.T

    return y_upd, P_upd, innov


def demo():
    np.random.seed(7)

    # --- Ground truth landmarks (unknown to the filter) ---
    landmarks = np.array([
        [4.0,  2.0],
        [8.0, -1.0],
        [2.0, -3.0],
    ], dtype=float)
    N = landmarks.shape[0]

    # --- True state ---
    x_true = np.array([0.0, 0.0, 0.0], dtype=float)

    # --- EKF-SLAM initial belief ---
    y = np.zeros(3 + 2 * N, dtype=float)
    y[:3] = np.array([0.0, 0.0, 0.0], dtype=float)

    # Initialize landmark guesses (poor prior)
    y[3:] = np.array([3.0,  1.0,  7.0,  0.0,  1.0, -2.0], dtype=float)

    P = np.eye(y.size, dtype=float) * 1e-3
    # Large uncertainty on map components
    for i in range(N):
        idx = 3 + 2 * i
        P[idx:idx + 2, idx:idx + 2] = np.eye(2) * 4.0

    # Noise models
    Q_pose = np.diag([0.02**2, 0.02**2, (np.deg2rad(1.0))**2])
    R_meas = np.diag([0.10**2, (np.deg2rad(2.0))**2])

    # Controls: (v, omega, dt)
    U = [
        np.array([1.0,  0.10, 1.0]),
        np.array([1.0,  0.00, 1.0]),
        np.array([1.0, -0.10, 1.0]),
        np.array([1.0,  0.00, 1.0]),
    ]

    for t, u in enumerate(U, start=1):
        # True motion with small perturbation
        x_true = motion_model(x_true, u) + np.random.multivariate_normal(np.zeros(3), Q_pose)
        x_true[2] = wrap_angle(x_true[2])

        # Predict
        y, P = ekf_slam_predict(y, P, u, Q_pose)

        # Observe all landmarks (known association), generate noisy measurements
        for i in range(N):
            z_true = measurement_model_pose_landmark(x_true, landmarks[i])
            z_noisy = z_true + np.random.multivariate_normal(np.zeros(2), R_meas)
            z_noisy[1] = wrap_angle(z_noisy[1])

            y, P, innov = ekf_slam_update_landmark(y, P, z_noisy, i, R_meas)

        print(f"t={t}: pose mean = {y[:3].round(3)}")

    print("\nFinal pose estimate:", y[:3])
    print("Final landmark estimates:")
    for i in range(N):
        idx = 3 + 2 * i
        print(f"  m{i}: {y[idx:idx+2]}")
    print("\nSmallest eigenvalue of P:", np.linalg.eigvalsh(P).min())


if __name__ == "__main__":
    demo()

File: Tutorials/Codes/Autonomous_Mobile_Robots_Course/Chapter11/Lesson1/Chapter11_Lesson1.m
% Chapter11_Lesson1.m
% Autonomous Mobile Robots (Control Engineering)
% Chapter 11: SLAM I — Filter-Based SLAM
% Lesson 1: The SLAM Problem Formulation
%
% This file contains:
%   (A) A minimal EKF-SLAM predict/update for an augmented state.
%   (B) A helper that can (optionally) generate a simple Simulink model shell
%       that calls the EKF step inside a MATLAB Function block.
%
% Note: EKF-SLAM algorithmic details are expanded in Lesson 2; here we use EKF
% only to operationalize the Bayes recursion for filter-based SLAM.

function Chapter11_Lesson1()
    rng(7);

    % --- Landmarks (ground truth, unknown to filter) ---
    landmarks = [4 2; 8 -1; 2 -3];
    N = size(landmarks,1);

    % --- True state ---
    x_true = [0;0;0];

    % --- EKF-SLAM belief ---
    y = zeros(3+2*N,1);
    y(1:3) = [0;0;0];                       % pose mean
    y(4:end) = [3;1; 7;0; 1;-2];            % rough landmark guesses

    P = eye(length(y))*1e-3;
    for i=1:N
        idx = 3 + 2*(i-1) + 1;
        P(idx:idx+1, idx:idx+1) = eye(2)*4.0;
    end

    Q_pose = diag([0.02^2, 0.02^2, (deg2rad(1.0))^2]);
    R_meas = diag([0.10^2, (deg2rad(2.0))^2]);

    U = [1.0  0.10 1.0;
         1.0  0.00 1.0;
         1.0 -0.10 1.0;
         1.0  0.00 1.0];

    for t=1:size(U,1)
        u = U(t,:).';

        % True motion with process noise
        x_true = motion_model(x_true, u) + mvnrnd([0 0 0], Q_pose).';
        x_true(3) = wrap_angle(x_true(3));

        % Predict
        [y,P] = ekf_slam_predict(y,P,u,Q_pose);

        % Observe all landmarks (known association)
        for i=1:N
            z_true  = meas_model(x_true, landmarks(i,:).');
            z_noisy = z_true + mvnrnd([0 0], R_meas).';
            z_noisy(2) = wrap_angle(z_noisy(2));

            [y,P] = ekf_slam_update_landmark(y,P,z_noisy,i,R_meas);
        end

        fprintf('t=%d: pose mean = [%.3f %.3f %.3f]\\n', t, y(1), y(2), y(3));
    end
end

function a = wrap_angle(a)
    a = mod(a + pi, 2*pi) - pi;
end

function x2 = motion_model(x,u)
    px=x(1); py=x(2); th=x(3);
    v=u(1); w=u(2); dt=u(3);

    if abs(w) < 1e-12
        px2 = px + v*dt*cos(th);
        py2 = py + v*dt*sin(th);
        th2 = th;
    else
        thN = th + w*dt;
        px2 = px + (v/w)*(sin(thN)-sin(th));
        py2 = py - (v/w)*(cos(thN)-cos(th));
        th2 = thN;
    end
    x2 = [px2; py2; wrap_angle(th2)];
end

function Fx = jac_motion_wrt_state(x,u)
    th=x(3);
    v=u(1); w=u(2); dt=u(3);
    Fx = eye(3);

    if abs(w) < 1e-12
        Fx(1,3) = -v*dt*sin(th);
        Fx(2,3) =  v*dt*cos(th);
    else
        thN = th + w*dt;
        Fx(1,3) = (v/w)*(cos(thN)-cos(th));
        Fx(2,3) = (v/w)*(sin(thN)-sin(th));
    end
end

function z = meas_model(x, m)
    px=x(1); py=x(2); th=x(3);
    dx = m(1)-px;
    dy = m(2)-py;
    r = sqrt(dx^2 + dy^2);
    b = atan2(dy,dx) - th;
    z = [r; wrap_angle(b)];
end

function [Hx,Hm] = jac_meas_wrt_pose_landmark(x,m)
    px=x(1); py=x(2);
    dx = m(1)-px;
    dy = m(2)-py;
    q = dx^2 + dy^2;
    if q < 1e-12, q = 1e-12; end
    r = sqrt(q);

    Hx = [ -dx/r, -dy/r, 0;
            dy/q, -dx/q, -1 ];
    Hm = [  dx/r,  dy/r;
           -dy/q,  dx/q ];
end

function [y_pred,P_pred] = ekf_slam_predict(y,P,u,Q_pose)
    n = length(y);
    x = y(1:3);
    Fx = jac_motion_wrt_state(x,u);
    x_pred = motion_model(x,u);

    F = eye(n);
    F(1:3,1:3) = Fx;

    Q = zeros(n);
    Q(1:3,1:3) = Q_pose;

    y_pred = y;
    y_pred(1:3) = x_pred;
    P_pred = F*P*F.' + Q;
end

function [y_upd,P_upd] = ekf_slam_update_landmark(y,P,z,landmark_id,R_meas)
    n = length(y);
    i = landmark_id - 1;     % 0-based
    idx = 3 + 2*i + 1;
    x = y(1:3);
    mi = y(idx:idx+1);

    z_hat = meas_model(x,mi);
    innov = [z(1)-z_hat(1); wrap_angle(z(2)-z_hat(2))];

    [Hx,Hm] = jac_meas_wrt_pose_landmark(x,mi);

    H = zeros(2,n);
    H(:,1:3) = Hx;
    H(:,idx:idx+1) = Hm;

    S = H*P*H.' + R_meas;
    K = P*H.'/S;

    y_upd = y + K*innov;
    y_upd(3) = wrap_angle(y_upd(3));

    I = eye(n);
    P_upd = (I-K*H)*P*(I-K*H).' + K*R_meas*K.';
end

% Optional: generate a tiny Simulink shell (requires Simulink)
function build_simulink_shell()
    mdl = 'SLAM_BayesFilter_Shell';
    if bdIsLoaded(mdl); close_system(mdl,0); end
    new_system(mdl); open_system(mdl);

    add_block('simulink/Sources/Constant', [mdl '/u'], 'Value', '[1;0.1;1]');
    add_block('simulink/Sources/Constant', [mdl '/z'], 'Value', '[5;0]');
    add_block('simulink/Sources/Constant', [mdl '/R'], 'Value', 'diag([0.1^2,(deg2rad(2))^2])');
    add_block('simulink/User-Defined Functions/MATLAB Function', [mdl '/EKFStep']);
    add_block('simulink/Sinks/Display', [mdl '/Display']);

    set_param([mdl '/EKFStep'], 'Script', ...
        ['function y_out = f(y_in,P_in,u,z,R)\n' ...
         '% Call an EKF update/predict in MATLAB code. This is a shell; wire in states as needed.\n' ...
         'y_out = y_in; %#ok<NASGU>\n' ...
         'end\n']);

    add_line(mdl,'u/1','EKFStep/3');
    add_line(mdl,'z/1','EKFStep/4');
    add_line(mdl,'R/1','EKFStep/5');
    add_line(mdl,'EKFStep/1','Display/1');

    save_system(mdl);
end
